x-pack/platform/plugins/shared/onechat/CONTRIBUTOR_GUIDE.md

```diff
@@ -0,0 +1,267 @@
+    };
+  },
+});
+```
+
+#### Tool result types
+
+For our framework to understand what kind of data is being returned by a tool, all tools
+must return a list of results following a specific format.
+
+- resource
+- tabular_data
+- query
+- other
+- error
+
+This is useful to allow the framework to perform specific processing on the results. For example,
+this is how we perform visualization rendering for the `tabular_data` type, by recognizing that
+a tool returned some result which can be rendered as a visualization if we want to.
+
+This is also how we render specific type of results differently in the UI, e.g we inline `query` results
+in the thinking panel.
+
+```ts
+onechat.tools.register({
+  id: 'platform.examples.result_types',
+  type: ToolType.builtin,
+  description: 'Some example',
+  tags: ['example'],
+  schema: z.object({
+    indexPattern: z.string().describe('Index pattern to filter on'),
+  }),
+  handler: async ({ indexPattern }, { events, esClient }) => {
+    const esqlQuery = await generateSomeQuery(indexPattern);
+    const data = await executeEsql(esqlQuery, esClient);
+
+    return {
+      results: [
+        { type: ToolResultType.query, data: { esql: esqlQuery } },
+        { type: ToolResultType.tabular_data, data },
+      ],
+    };
+  },
+});
+```
+
+See the `ToolResultType` and corresponding types in `x-pack/platform/packages/shared/onechat/onechat-common/tools/tool_result.ts`
+
+### Registering other types of tools
+
+Platform contributors aren't stuck to using the `builtin` tool type. They are free to leverage the other
+existing tool types, and create static instances of them.
+
+E.g. registering a built-in `index_search` tool:
+
+```ts
+onechatSetup.tools.register({
+  id: 'platform.core.some_knowledge_base',
+  type: ToolType.index_search,
+  description: 'Use this tool to retrieve documentation from our knowledge base',
+  configuration: {
+    pattern: '.my_knowledge_base',
+  },
+});
+```
+
+## Registering built-in agents
+
+### Registering the agent
+
+```ts
+class MyPlugin {
+  setup(core: CoreSetup, { onechat }: { onechat: OnechatPluginSetup }) {
+    onechat.agents.register(myAgentDefinition);
+  }
+}
+```
+
+### Adding the agent to the allow list
+
+Similar to tools, we keep an hardcoded list of registered agents to trigger a code review from the team when
+agents are added.
+
+To add a tool to the allow list, simply add the tool's id to the `AGENT_BUILDER_BUILTIN_AGENTS` array,
+in `x-pack/platform/packages/shared/onechat/onechat-server/allow_lists.ts`
+
+(Kibana will fail to start otherwise, with an explicit error message explaining what to do)
+
+### Basic example
+
+How registering a basic agent looks like:
+
+```ts
+onechat.agents.register({
+  id: 'platform.core.dashboard',
+  name: 'Dashboard agent',
+  description: 'Agent specialized in dashboard related tasks',
+  avatar_icon: 'dashboardApp',
+  configuration: {
+    instructions: 'You are a dashboard specialist [...]',
```

Comment:
instructions can also be a handler too right? so they can have access to the model provider

Reply:
Not yet - we decided to KISS and keep the configuration the exact same for the time being (especially given we were told from security they might not need it that much after all)